In [2]:
# Libraries import karo
import pandas as pd
import numpy as np
import yfinance as yf

# Data dobara load karo
df = yf.download("RELIANCE.NS", start="2020-01-01", end="2026-04-05")

print("Data loaded successfully! ✅")
print("Shape:", df.shape)

[*********************100%***********************]  1 of 1 completed

1 Failed download:
['RELIANCE.NS']: TypeError("'NoneType' object is not subscriptable")


Data loaded successfully! ✅
Shape: (0, 6)


In [3]:
import pandas as pd
import numpy as np
import yfinance as yf

# Thoda alag tarike se download karo
ticker = yf.Ticker("RELIANCE.NS")
df = ticker.history(start="2020-01-01", end="2026-04-05")

print("Data loaded successfully! ✅")
print("Shape:", df.shape)
df.head()

Data loaded successfully! ✅
Shape: (1549, 7)


,Open,High,Low,Close,Volume,Dividends,Stock Splits
Date,,,,,,,
2020-01-01 00:00:00+05:30,679.081936,683.152852,673.490062,675.324158,14004468,0.0,0.0
2020-01-02 00:00:00+05:30,676.397899,689.348791,676.397899,686.821228,17710316,0.0,0.0
2020-01-03 00:00:00+05:30,685.792313,689.661956,681.318790,687.648865,20984698,0.0,0.0
2020-01-06 00:00:00+05:30,679.976719,683.510767,670.134933,671.700684,24519177,0.0,0.0
2020-01-07 00:00:00+05:30,679.529321,686.463335,677.068889,682.034546,16683622,0.0,0.0


In [4]:
df.tail()

,Open,High,Low,Close,Volume,Dividends,Stock Splits
Date,,,,,,,
2026-03-25 00:00:00+05:30,1420.000000,1430.500000,1408.400024,1413.099976,19041331,0.0,0.0
2026-03-27 00:00:00+05:30,1401.000000,1404.699951,1345.000000,1348.099976,23266517,0.0,0.0
2026-03-30 00:00:00+05:30,1335.000000,1365.000000,1334.099976,1343.900024,24393861,0.0,0.0
2026-04-01 00:00:00+05:30,1384.199951,1384.400024,1362.900024,1369.199951,14404619,0.0,0.0
2026-04-02 00:00:00+05:30,1357.000000,1358.199951,1328.000000,1350.500000,21280051,0.0,0.0


In [5]:
# Missing values check karo
print("Missing values har column mein:")
print(df.isnull().sum())

Missing values har column mein:
Open            0
High            0
Low             0
Close           0
Volume          0
Dividends       0
Stock Splits    0
dtype: int64


In [6]:
# Duplicate rows check karo
print("Duplicate rows:", df.duplicated().sum())

Duplicate rows: 0


In [7]:
# Har column ka data type dekho
print("Data types:")
print(df.dtypes)

Data types:
Open            float64
High            float64
Low             float64
Close           float64
Volume            int64
Dividends       float64
Stock Splits    float64
dtype: object


In [10]:
# Timezone hatao date se
df.index = df.index.tz_localize(None)

# Confirm karo
print("Date fixed:")
print(df.index[:10])

# Baad mein jab hum ML model banayenge toh date se naye features nikalenge jaise month, year, day of week etc. Agar timezone laga hota toh wahan error aata! Isliye abhi hi fix kar liya. ✅

Date fixed:
DatetimeIndex(['2020-01-01', '2020-01-02', '2020-01-03', '2020-01-06',
               '2020-01-07', '2020-01-08', '2020-01-09', '2020-01-10',
               '2020-01-13', '2020-01-14'],
              dtype='datetime64[ns]', name='Date', freq=None)


In [11]:
# Unnecessary columns drop karo
df = df.drop(columns=["Dividends", "Stock Splits"])

# Confirm karo
print("Columns after dropping:")
print(df.columns)
print("New shape:", df.shape)

Columns after dropping:
Index(['Open', 'High', 'Low', 'Close', 'Volume'], dtype='object')
New shape: (1549, 5)


In [12]:
df.head()

,Open,High,Low,Close,Volume
Date,,,,,
2020-01-01,679.081936,683.152852,673.490062,675.324158,14004468
2020-01-02,676.397899,689.348791,676.397899,686.821228,17710316
2020-01-03,685.792313,689.661956,681.318790,687.648865,20984698
2020-01-06,679.976719,683.510767,670.134933,671.700684,24519177
2020-01-07,679.529321,686.463335,677.068889,682.034546,16683622


In [14]:
# Poore data ka summary dekho
print("Dataset Info:")
print("=============")
print(f"Total Trading Days: {df.shape[0]}")
print(f"Total Columns: {df.shape[1]}")
print(f"Date Range: {df.index[0].date()} to {df.index[-1].date()}")
print(f"\nBasic Statistics:")
df.describe()



# Min price ₹395 — ye COVID-19 crash (March 2020) ka time tha jab pure market ne gira tha! Aur tab se price 4x ho gayi ₹1592 tak! Ye ek powerful business insight hai jo tum resume mein mention kar sakte ho! 😎

Dataset Info:
Total Trading Days: 1549
Total Columns: 5
Date Range: 2020-01-01 to 2026-04-02

Basic Statistics:


,Open,High,Low,Close,Volume
count,1549.000000,1549.000000,1549.000000,1549.000000,1.549000e+03
mean,1159.311317,1170.538465,1147.549024,1158.573798,1.799641e+07
std,241.935895,242.250156,241.747023,242.058944,1.493154e+07
min,409.864893,424.985445,391.724735,395.482513,0.000000e+00
25%,1009.610086,1021.456482,998.296943,1007.009827,9.575812e+06
50%,1167.741314,1179.587432,1159.074855,1168.813965,1.333409e+07
75%,1374.000000,1384.000000,1360.430819,1371.721436,2.038580e+07
max,1593.000000,1611.800049,1578.199951,1592.300049,1.426834e+08


# Day 2 - Data Cleaning
- No missing values found
- No duplicate rows found
- Data types verified — float64 and int64
- Timezone removed from Date index
- Dropped unnecessary columns: Dividends, Stock Splits
- Final dataset: 1549 rows, 5 columns
- Date range: 2020-01-01 to 2026-04-02